File: samples/sample.py
#!/usr/bin/env python3
"""
Void Space Theme - Python Sample
Demonstrates: types, functions, classes, decorators, comprehensions, exceptions.
"""

from __future__ import annotations

import os
import sys
import math
import asyncio
from dataclasses import dataclass, field
from typing import Optional, Union, Generator
from enum import Enum, auto


# Constants

VERSION: str = "1.0.0"
MAX_RETRIES: int = 3
PI: float = math.pi
EMPTY: None = None

PALETTE: dict[str, str] = {
    "void":    "#0d0f18",
    "nebula":  "#c792ea",
    "aurora":  "#89ddff",
    "comet":   "#82aaff",
}


# Enum

class Planet(Enum):
    MERCURY = auto()
    VENUS   = auto()
    EARTH   = auto()
    MARS    = auto()

    def distance_from_sun(self) -> float:
        """Returns approximate distance in AU."""
        distances = {
            Planet.MERCURY: 0.39,
            Planet.VENUS:   0.72,
            Planet.EARTH:   1.00,
            Planet.MARS:    1.52,
        }
        return distances[self]


# Dataclass

@dataclass
class Star:
    name: str
    mass: float                        # solar masses
    luminosity: float                  # solar luminosities
    tags: list[str] = field(default_factory=list)
    _cache: dict = field(default_factory=dict, repr=False)

    def __post_init__(self) -> None:
        if self.mass <= 0:
            raise ValueError(f"Mass must be positive, got {self.mass}")

    @property
    def spectral_class(self) -> str:
        if self.mass >= 16:   return "O"
        if self.mass >= 2.1:  return "B"
        if self.mass >= 1.4:  return "A"
        if self.mass >= 1.04: return "F"
        if self.mass >= 0.8:  return "G"
        if self.mass >= 0.45: return "K"
        return "M"

    def __repr__(self) -> str:
        return f"Star(name={self.name!r}, class={self.spectral_class})"


# Decorator

def retry(times: int = 3):
    """Decorator that retries a function on exception."""
    def decorator(fn):
        def wrapper(*args, **kwargs):
            last_exc: Optional[Exception] = None
            for attempt in range(times):
                try:
                    return fn(*args, **kwargs)
                except Exception as exc:
                    last_exc = exc
                    print(f"[retry] attempt {attempt + 1}/{times} failed: {exc}")
            raise RuntimeError(f"All {times} attempts failed") from last_exc
        return wrapper
    return decorator


@retry(times=MAX_RETRIES)
def fetch_star_data(star_id: int) -> dict:
    if star_id < 0:
        raise ValueError(f"Invalid star_id: {star_id}")
    return {"id": star_id, "name": "Sirius", "mass": 2.1}


# Generator / comprehension

def fibonacci() -> Generator[int, None, None]:
    a, b = 0, 1
    while True:
        yield a
        a, b = b, a + b


def first_n_fibs(n: int) -> list[int]:
    gen = fibonacci()
    return [next(gen) for _ in range(n)]


squares = {x: x**2 for x in range(1, 11)}
evens   = [n for n in range(100) if n % 2 == 0]
matrix  = [[row * col for col in range(1, 4)] for row in range(1, 4)]


# Async

async def observe(star: Star, duration: float) -> str:
    await asyncio.sleep(duration)
    return f"Observed {star.name} for {duration:.2f}s — class {star.spectral_class}"


async def survey(stars: list[Star]) -> list[str]:
    tasks = [observe(s, 0.01) for s in stars]
    return await asyncio.gather(*tasks)


# Context manager

class Telescope:
    def __init__(self, aperture_mm: int) -> None:
        self.aperture = aperture_mm
        self._open = False

    def __enter__(self) -> "Telescope":
        self._open = True
        print(f"Telescope ({self.aperture}mm) opened")
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> bool:
        self._open = False
        print("Telescope closed")
        return False   # do not suppress exceptions

    def observe(self, target: str) -> None:
        if not self._open:
            raise RuntimeError("Telescope is closed")
        print(f"Observing: {target}")


# Union / match

def describe(obj: Union[Star, Planet, str]) -> str:
    match obj:
        case Star(name=n, mass=m) if m > 10:
            return f"{n} is a massive star"
        case Star(name=n):
            return f"{n} is an ordinary star"
        case Planet.EARTH:
            return "Home planet"
        case Planet():
            return f"Planet: {obj.name}"
        case str(s):
            return f"Unknown: {s}"
        case _:
            return "Unrecognised object"


# Entry point

def main() -> int:
    sirius  = Star("Sirius",  mass=2.10, luminosity=25.4, tags=["bright", "binary"])
    betelgeuse = Star("Betelgeuse", mass=20.0, luminosity=1e5)

    stars = [sirius, betelgeuse]

    print(f"void-space v{VERSION}")
    print(f"Pi ≈ {PI:.6f}")
    print(f"Fibonacci(10): {first_n_fibs(10)}")
    print(f"Squares: {squares}")

    for s in stars:
        print(describe(s))

    print(f"Earth is {Planet.EARTH.distance_from_sun()} AU from the Sun")

    with Telescope(aperture_mm=200) as scope:
        scope.observe("Andromeda Galaxy")

    results = asyncio.run(survey(stars))
    for r in results:
        print(r)

    env_home = os.environ.get("HOME", "/tmp")
    print(f"Running on {sys.platform}, HOME={env_home}")

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: samples/sample.js
/**
 * Void Space Theme — JavaScript Sample
 * Demonstrates: classes, async/await, closures, generators, destructuring,
 * template literals, symbols, WeakMap, Proxy, and modern JS patterns.
 */

'use strict';

// Constants

const VERSION = '1.0.0';
const MAX_RETRIES = 3;
const PI = Math.PI;
const NONE = null;

/** @type {Record<string, string>} */
const PALETTE = Object.freeze({
  void:   '#0d0f18',
  nebula: '#c792ea',
  aurora: '#89ddff',
  comet:  '#82aaff',
  pulsar: '#ff5370',
});

const SYM_PRIVATE = Symbol('private');


// Enum-like object

const SpectralClass = Object.freeze({
  O: 'O', B: 'B', A: 'A', F: 'F', G: 'G', K: 'K', M: 'M',
});


// Class hierarchy

class CelestialBody {
  #name;

  constructor(name) {
    if (new.target === CelestialBody) {
      throw new TypeError('CelestialBody is abstract');
    }
    this.#name = String(name);
  }

  get name() { return this.#name; }

  describe() {
    throw new Error(`${this.constructor.name}.describe() not implemented`);
  }

  toString() { return `[${this.constructor.name} "${this.name}"]`; }
}


class Star extends CelestialBody {
  #observations = 0;
  [SYM_PRIVATE] = {};

  /**
   * @param {string}   name
   * @param {number}   massSolar      - solar masses
   * @param {number}   luminosity     - solar luminosities
   * @param {string[]} [tags]
   */
  constructor(name, massSolar, luminosity, tags = []) {
    super(name);
    if (massSolar <= 0) throw new RangeError(`mass must be positive, got ${massSolar}`);
    this.massSolar   = massSolar;
    this.luminosity  = luminosity;
    this.tags        = [...tags];
  }

  get spectralClass() {
    const m = this.massSolar;
    if (m >= 16)   return SpectralClass.O;
    if (m >= 2.1)  return SpectralClass.B;
    if (m >= 1.4)  return SpectralClass.A;
    if (m >= 1.04) return SpectralClass.F;
    if (m >= 0.8)  return SpectralClass.G;
    if (m >= 0.45) return SpectralClass.K;
    return SpectralClass.M;
  }

  observe() { this.#observations++; }
  get observationCount() { return this.#observations; }

  describe() {
    return `${this.name} [class ${this.spectralClass}, ` +
           `mass=${this.massSolar.toFixed(2)}☉, lum=${this.luminosity.toExponential(2)}☉]`;
  }

  toJSON() {
    return { name: this.name, spectralClass: this.spectralClass,
             massSolar: this.massSolar, luminosity: this.luminosity, tags: this.tags };
  }
}


// Mixin pattern

const Serialisable = (Base) => class extends Base {
  serialise() { return JSON.stringify(this); }

  static deserialise(json) {
    const { name, massSolar, luminosity, tags } = JSON.parse(json);
    return new this(name, massSolar, luminosity, tags);
  }
};

class SerialisableStar extends Serialisable(Star) {}


// Decorator (stage-3 via manual wrapping)

function retry(times = 3) {
  return function (fn) {
    return async function (...args) {
      let lastErr;
      for (let attempt = 0; attempt < times; attempt++) {
        try {
          return await fn.apply(this, args);
        } catch (err) {
          lastErr = err;
          console.warn(`[retry] attempt ${attempt + 1}/${times} failed:`, err.message);
        }
      }
      throw new Error(`All ${times} attempts failed`, { cause: lastErr });
    };
  };
}

const fetchStarData = retry(MAX_RETRIES)(async function (starId) {
  if (typeof starId !== 'number' || starId < 0) {
    throw new TypeError(`Invalid starId: ${starId}`);
  }
  // Simulated async fetch
  await delay(10);
  return { id: starId, name: 'Sirius', massSolar: 2.1 };
});


// Async / Promise

function delay(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function observeStar(star, durationMs = 50) {
  await delay(durationMs);
  star.observe();
  return `Observed ${star.name} — ${star.observationCount} observation(s)`;
}

async function surveyAll(stars) {
  const results = await Promise.all(stars.map((s) => observeStar(s)));
  return results;
}


// Generator

function* fibonacci() {
  let [a, b] = [0, 1];
  while (true) {
    yield a;
    [a, b] = [b, a + b];
  }
}

function firstN(gen, n) {
  const out = [];
  for (const val of gen) {
    out.push(val);
    if (out.length >= n) break;
  }
  return out;
}


// Proxy

function makeReadonly(obj) {
  return new Proxy(obj, {
    set(_, prop) {
      throw new TypeError(`Cannot set property "${prop}" on a read-only object`);
    },
    deleteProperty(_, prop) {
      throw new TypeError(`Cannot delete property "${prop}" on a read-only object`);
    },
  });
}


// Closure / IIFE

const makeCounter = (start = 0) => {
  let n = start;
  return {
    next:  () => ++n,
    reset: () => { n = start; },
    value: () => n,
  };
};

const catalogueSize = (() => {
  const sizes = [1024, 2048, 4096];
  return sizes.reduce((acc, s) => acc + s, 0);
})();


// Destructuring & spread

const [primary, ...rest] = ['void', 'nebula', 'aurora', 'comet'];
const { void: bg, nebula: accent, ...others } = PALETTE;

function formatStar({ name, massSolar, luminosity, tags = [] }) {
  return `${name} (${massSolar}☉) [${tags.join(', ')}]`;
}


// Tagged template literal

function highlight(strings, ...values) {
  return strings.reduce((acc, str, i) => {
    const val = values[i] !== undefined ? `\x1b[33m${values[i]}\x1b[0m` : '';
    return acc + str + val;
  }, '');
}


// WeakMap (private data pattern)

const _secrets = new WeakMap();

class Observatory {
  constructor(location, apertureMm) {
    _secrets.set(this, { location, apertureMm, log: [] });
  }

  record(entry) {
    _secrets.get(this).log.push({ ts: Date.now(), entry });
  }

  report() {
    const { location, apertureMm, log } = _secrets.get(this);
    return { location, apertureMm, entries: log.length };
  }
}


// Error subclass

class ObservationError extends Error {
  constructor(starName, cause) {
    super(`Observation of "${starName}" failed`);
    this.name      = 'ObservationError';
    this.starName  = starName;
    this.cause     = cause;
  }
}


// Entry point

(async () => {
  console.log(`void-space v${VERSION} — Pi ≈ ${PI.toFixed(6)}`);

  const sirius     = new Star('Sirius',           2.10,  25.4,  ['bright', 'binary']);
  const betelgeuse = new Star('Betelgeuse',       20.0,  1e5,   ['red-supergiant']);
  const proxima    = new Star('Proxima Centauri', 0.12,  0.0017);

  const stars = [sirius, betelgeuse, proxima];

  stars.forEach((s) => console.log(s.describe()));

  // Async survey
  const observations = await surveyAll(stars);
  observations.forEach((r) => console.log(' >', r));

  // Fetch with retry
  const data = await fetchStarData(42);
  console.log('Fetched:', data);

  // Generator
  const fibs = firstN(fibonacci(), 10);
  console.log('Fibonacci:', fibs);

  // Closure counter
  const counter = makeCounter(0);
  [1, 2, 3, 4, 5].forEach(() => console.log('tick', counter.next()));

  // Tagged template
  console.log(highlight`Star ${sirius.name} has class ${sirius.spectralClass}`);

  // Destructuring
  console.log(`Primary colour key: ${primary}, rest: ${rest.join(', ')}`);
  console.log(`bg=${bg}, accent=${accent}`);
  console.log(formatStar({ name: 'Sirius', massSolar: 2.1, luminosity: 25, tags: ['bright'] }));

  // Proxy readonly
  const ro = makeReadonly({ answer: 42 });
  try { ro.answer = 0; } catch (e) { console.log('readonly error:', e.message); }

  // Observatory
  const obs = new Observatory('Mauna Kea', 10000);
  stars.forEach((s) => obs.record(s.describe()));
  console.log(obs.report());

  // Error subclass
  const obsErr = new ObservationError('Ghost', new Error('not found'));
  console.log(obsErr.message, '—', obsErr instanceof Error);

  console.log(`catalogueSize=${catalogueSize}`);
  console.log(`Serialised: ${new SerialisableStar('Sol', 1, 1).serialise()}`);
})();
